### PR TITLE
feat: fp-1736 allow form to send email

### PR DESCRIPTION
**Pull request by wesleyboar**
## Overview

Allow forms to send an e-mail.

_Remote server needs access to an SMTP server. Direction I can provide is to override settings for specific server (like APCD). Reference: [Django docs for email settings](https://docs.djangoproject.com/en/2.1/ref/settings/#email-backend)._

## Related

- [FP-1736](https://jira.tacc.utexas.edu/browse/FP-1736)
- required by https://github.com/TACC/Core-CMS-Custom/pull/4

## Changes

- add e-mail settings
- note another known issue with forms (not related to e-mail)
- update form fields to show email feature
- add [developer guide for cms e-mail](https://confluence.tacc.utexas.edu/x/coR9E)
- import e-mail settings

### Remote

<details><summary>Run and test production and staging e-mail settings and TACC mail server.</summary>

1. Be an admin on https://apcd-qa.tacc.utexas.edu/.
2. Open https://apcd-qa.tacc.utexas.edu/forms/fp-1773-e-mail-test/.

#### Default Sender

3. Edit form:
    - "SEND FORM DATA TO E-MAIL ADDRESS" is an e-mail address you can check.
    - "SENDER EMAIL ADDRESS" is blank.
4. Save form.
5. Submit form.
6. Verify receipt of e-mail.

#### Custom Sender

7. Edit form:
    - "SEND FORM DATA TO E-MAIL ADDRESS" is an e-mail address you can check.
    - "SENDER EMAIL ADDRESS" is`no-reply@apcd-qa.tacc.utexas.edu`.
8. Save form.
9. Submit form.
10. Verify receipt of e-mail.

</details>

### Local

<details><summary>Run and test development e-mail settings and server.</summary>

0. Have a running CMS.
1. On local machine, `docker exec -it core_cms /bin/bash` (or `… sad_cms …` if you [have that setup](https://github.com/TACC/Core-CMS/wiki/Locally-Develop-CMS---Portal---Docs))
8. In docker container, `python -m smtpd -n -c DebuggingServer localhost:1025`
9. On running CMS, create and submit e-mail–sending form:
    1. Login.
    2. As an admin, create a form.
    3. While creating form, set up form to send e-mail (might be the default) (see "Submission settings").
    10. As a user, submit the form.
5. Check stdout from container (e.g. via Docker Desktop).
11. Verify log of submitted form and its data.

    <details><summary>example output</summary>

    ```bash
    [12/Aug/2022 13:35:12] "POST /forms/submit/ HTTP/1.1" 500 27268
    Content-Type: multipart/alternative;
     boundary="===============1722349033141855018=="
    MIME-Version: 1.0
    Subject: Wes' Localhost Test From
    From: nobody@tacc.utexas.edu
    To: wbomar@tacc.utexas.edu
    Date: Fri, 12 Aug 2022 18:35:41 -0000
    Message-ID: <166032934101.58.13019985486494731103@sad_cms>

    --===============1722349033141855018==
    Content-Type: text/plain; charset="utf-8"
    MIME-Version: 1.0
    Content-Transfer-Encoding: 8bit



    Form: Test Form 1
    Date: 12th August 2022 13:35


    ...
    ```
    </details>

</details>

## UI

| debug log | cms form | email receipt |
| - | - | - |
| stdout form email send | Screen Shot 2022-08-12 at 14 11 19 | Screen Shot 2022-08-15 at 12 52 18 |

### CODE REVIEW

File: taccsite_cms/_settings/form_plugin.py [hunk 1]
```diff
@@ -67,26 +75,24 @@
     (
         _('Submission settings'),
         {
-            'description': 'Whether to save form data.',
+            'description': 'You can choose to set from which, and to which, e-mail address to send form submissions and/or to save submitted form data in the CMS database. <small>The "SENDER_EMAIL_ADDRESS" may remain blank, so server default no-reply address would be used.</small>',
```
Comment (wesleyboar) [inline taccsite_cms/_settings/form_plugin.py:78]: This is the default text, with `<small>` text I appended to explain why one may want `email_from` not set.

File: taccsite_cms/_settings/email.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+########################
+# DJANGO
+# https://docs.djangoproject.com/en/2.2/topics/email/
+########################
+
+# Development, Staging, Production
+# https://confluence.tacc.utexas.edu/x/coR9E
+# EMAIL_HOST = "relay.tacc.utexas.edu"
+# DEFAULT_FROM_EMAIL = 'no-reply@...' # where ... is project's production domain
```
Comment (wesleyboar) [inline taccsite_cms/_settings/email.py:11]: I added these to APCD settings in https://github.com/TACC/Core-CMS-Custom/pull/4.

File: taccsite_cms/_settings/email.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+# Debugging
+# https://docs.djangoproject.com/en/2.1/topics/email/#configuring-email-for-development
+# EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
+# EMAIL_HOST = 'localhost'
+# EMAIL_PORT = 1025
+# DEFAULT_FROM_EMAIL = 'no-reply@localhost'
```
Comment (wesleyboar) [inline taccsite_cms/_settings/email.py:18]: I tested these locally. They worked. (I could see email content printed to container console output.)

**Review by wesleyboar — COMMENTED** [review]
Notes for reviewers.

File: taccsite_cms/_settings/form_plugin.py [hunk 4]
```diff
@@ -10,6 +10,14 @@
 # (No success hiding those formats from editor)
 # https://github.com/avryhof/djangocms-forms/issues/8
 
+# Unable to submit files
+# > AttributeError at /forms/submit/
+# ```
+#   File "/opt/pysetup/.venv/src/djangocms-forms-maintained/djangocms_forms/fields.py", line 42, in clean
+#     if uploaded_file._size > self.max_upload_size:
+# AttributeError: 'InMemoryUploadedFile' object has no attribute '_size'
+# ```
```
Comment (wesleyboar) [inline taccsite_cms/_settings/form_plugin.py:19]: This is not related to e-mail. I just noticed thus noted another known issue with the plugin.

Comment (rstijerina) [reply taccsite_cms/_settings/form_plugin.py:19]: Is this the output for a file attachment attempt?

File: taccsite_cms/_settings/form_plugin.py (hunk 1 above)
Comment (rstijerina) [reply taccsite_cms/_settings/form_plugin.py:78]: Could maybe be worded as
```
<small>Portal default from address (DEFAULT_FROM_EMAIL) will be used if "SENDER_EMAIL_ADDRESS" is blank.</small>
```

**Review by rstijerina — APPROVED** [review]
LGTM. Left some comments

File: taccsite_cms/_settings/form_plugin.py (hunk 4 above)
Comment (wesleyboar) [reply taccsite_cms/_settings/form_plugin.py:19]: Yes.

Seems simple enough to fix. Probably some DJANGO setting. But the image I tested was small, barely over 500KB I think. I haven't tried to fix this yet.

Comment (rstijerina) [reply taccsite_cms/_settings/form_plugin.py:19]: Maybe we can open an issue with their library and link that here, then we could remove this commented stacktrace

File: taccsite_cms/_settings/form_plugin.py (hunk 1 above)
Comment (wesleyboar) [reply taccsite_cms/_settings/form_plugin.py:78]: Yes, better.

Okay. And I'll tweak it to not assume a portal, and fix my original typo, and add some formatting.

```suggestion
            'description': 'You can choose to set from which, and to which, e-mail address to send form submissions and/or to save submitted form data in the CMS database. <small>The CMS default <em>from</em> address (<code>DEFAULT_FROM_EMAIL</code>) will be used if "SENDER EMAIL ADDRESS" is blank.</small>',
```

File: taccsite_cms/_settings/form_plugin.py (hunk 4 above)
Comment (wesleyboar) [reply taccsite_cms/_settings/form_plugin.py:19]:
```suggestion
# Unable to submit files
# https://github.com/avryhof/djangocms-forms/issues/14
```
